geopandas/base.py

```diff
@@ -1008,9 +1008,55 @@ def buffer(self, distance, resolution=16, **kwargs):
         distance : float, np.array, pd.Series
             The radius of the buffer. If np.array or pd.Series are used
             then it must have same length as the GeoSeries.
-        resolution: int
-            Optional, the resolution of the buffer around each vertex.
+        resolution : int (optional, default 16)
+            The resolution of the buffer around each vertex.
+
+        Examples
+        --------
+        >>> from shapely.geometry import Point, LineString, Polygon
+        >>> s = geopandas.GeoSeries(
+        ...             [
+        ...                 Point(0, 0),
+        ...                 LineString([(1, -1), (1, 0), (2, 0), (2, 1)]),
+        ...                 Polygon([(3, -1), (4, 0), (3, 1)]),
+        ...             ]
+        ...         )
+        >>> s
+        0                              POINT (0.00000 0.00000)
+        1    LINESTRING (1.00000 -1.00000, 1.00000 0.00000,...
+        2    POLYGON ((3.00000 -1.00000, 4.00000 0.00000, 3...
+        dtype: geometry
+
+        >>> s.buffer(0.2)
+        0    POLYGON ((0.20000 0.00000, 0.19904 -0.01960, 0...
+        1    POLYGON ((0.80000 0.00000, 0.80096 0.01960, 0....
+        2    POLYGON ((2.80000 -1.00000, 2.80000 1.00000, 2...
+        dtype: geometry
+
+        ``**kwargs`` accept further specification as ``join_style`` and ``cap_style``.
+        See the following illustration of different options.
+
+        >>> import matplotlib.pyplot as plt
+        >>> fix, axs = plt.subplots(3, 2, figsize=(12, 12), sharex=True, sharey=True)
+        >>> for ax in axs.flatten():  # doctest: +SKIP
+        ...     s.plot(ax=ax)
+        >>> s.buffer(0.2).plot(ax=axs[0,0], alpha=.6)  # doctest: +SKIP
+        >>> axs[0,0].set_title('s.buffer(0.2)')  # doctest: +SKIP
+        >>> s.buffer(0.2, resolution=2).plot(ax=axs[0,1], alpha=.6)  # doctest: +SKIP
+        >>> axs[0,1].set_title('s.buffer(0.2, resolution=2)')  # doctest: +SKIP
+        >>> s.buffer(0.2, cap_style=2).plot(ax=axs[1,0], alpha=.6)  # doctest: +SKIP
+        >>> axs[1,0].set_title('s.buffer(0.2, cap_style=2)')  # doctest: +SKIP
+        >>> s.buffer(0.2, cap_style=3).plot(ax=axs[1,1], alpha=.6)  # doctest: +SKIP
+        >>> axs[1,1].set_title('s.buffer(0.2, cap_style=3)')  # doctest: +SKIP
+        >>> s.buffer(0.2, join_style=2).plot(ax=axs[2,0], alpha=.6)  # doctest: +SKIP
+        >>> axs[2,0].set_title('s.buffer(0.2, join_style=2)')  # doctest: +SKIP
+        >>> s.buffer(0.2, join_style=3).plot(ax=axs[2,1], alpha=.6)  # doctest: +SKIP
+        >>> axs[2,1].set_title('s.buffer(0.2, join_style=3)')  # doctest: +SKIP
+
+        .. image:: ../../../_static/buffer_illustrations.svg
+            :align: center
```

Comment:
Again, some comments that could also be for a follow-up: it's a great illustration to add here! And this is something we should start doing more for other docstrings as well. 

But, when doing this more generally, we should maybe think about a more streamlined/generalized workflow for this. Some first ideas:

- Put the code to generate the image in separate files (eg one per image). This might make it more maintainable (eg easier to do few edits, and re-generate all images). We could from here still include both the generated image as the code (sphinx can include a code snippet from a file, if we want, we could maybe also opt to *not* show the code)
- Use the matplotlib `.. plot::` directive (https://matplotlib.org/3.1.1/devel/plot_directive.html) to include them. I think this can both work with the code in separate scripts (this is what shapely does, see eg https://github.com/Toblerity/Shapely/blame/master/docs/manual.rst#L345) as with the code in the docstring itself (pandas does this in the plot() docstrings, eg https://github.com/pandas-dev/pandas/blob/7ffbf1a5ab6b0126c7efa5991e7db505979a752c/pandas/plotting/_core.py#L201-L208). 
  With this option, the figures would be created on each doc build. This could ensure automatically that the generated figures are never out of date, but of course makes the doc building slower.

(but again, we should maybe discuss this in a separate issue)

Reply:
Good points! I like the idea of autogenerated plots, I'll give it a go in this PR, maybe combined with an external file for the code.